### PR TITLE
[gdscript/en] Update for Godot 4

### LINKED ISSUE #4470: [Request] [GDScript2.0/en]
There's already a GDScript guide here, but Godot 4.0 is close to beta and it introduces GDScript 2.0, which has many compatibility breaking features. Maybe a new guide should be added (like with Python2 and Python3) or at least parts of the old one should be rewritten. Anyway here are some resources:
- https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_basics.html
- https://gamefromscratch.com/gdscript-2-in-godot-4/
- https://godotengine.org/article/gdscript-progress-report-feature-complete-40
- https://godotengine.org/article/gdscript-progress-report-new-gdscript-now-merged

### PR DESCRIPTION
**Pull request by zacryol**
resolves #4470 

- [x] I solemnly swear that this is all original content of which I am the original author
  - (It's based on the original GDScript section, so not entirely mine)
- [x] Pull request title is prepended with `[language/lang-code]` (example `[python/fr-fr]` or `[java/en]`)
- [x] Pull request touches only one file (or a set of logically related files with similar changes made)
- [x] Content changes are aimed at *intermediate to experienced programmers* (this is a poor format for explaining fundamental programming concepts)
- [x] If you've changed any part of the YAML Frontmatter, make sure it is formatted according to [CONTRIBUTING.md](https://github.com/adambard/learnxinyminutes-docs/blob/master/CONTRIBUTING.markdown)
  - [x] Yes, I have double-checked quotes and field names!

Should cover:
 - [x] use of `@onready`
 - [x] use of various `@export` annotations
   - [x] base `@export`
   - [x] ~~more specific annotations~~ Probably not needed for a basic introduction.
 - [x] Documentation Comments
 - [x] enums as types
 - [x] Packed Arrays instead of Pool
 - [x] New NodePath syntax (`^` instead of `@`)
 - [x] Directly using a Node as an export type
 - [x] Remove the use of a Tween node
 - [x] Typed signals
 - [x] Typed Arrays
 - [x] Lambda expressions
 - [x] More emphasis on static typing (more of an opinion)
   - Added stuff to the section
 - [x] Change superclass method call to `super()`
 - [ ] Anything else that people may bring up.

### CODE REVIEW

**Comment by Calinou** [comment]
Should it be a separate page? I think we should aim to have information for both GDScript versions on the same page.

PS: We don't formally refer to GDScript in Godot 4.0 as "GDScript 2.0", so I'd avoid that terminology as it can be confusing.

File: gdscript2.html.markdown [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+    - ["Wichamir", "https://github.com/Wichamir/"]
+    - ["zacryol", "https://github.com/zacryol"]
+filename: learngdscript2.gd
+---
+
+GDScript 2.0 is a dynamically and statically typed scripting language
+made specifically for free and open source game engine Godot, now revamped
+for the new Godot Engine 4.0. GDScript's syntax is vaguely similar to
+Python's. Its main advantages are ease of use and tight integration with
+the engine. It's a perfect fit for game development.
```
Comment (superherointj) [inline gdscript2.html.markdown:13]:
```suggestion
```
Shorter is better.

File: gdscript2.html.markdown [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+extends Node2D
+
+# Member variables
+var x = 8 # int
+var y = 1.2 # float
+var b = true # bool
+var s = "Hello World!" # String
+var a = [1, false, "brown fox"] # Array - similar to list in Python,
+                                # it can hold different types
+                                # of variables at once.
```
Comment (superherointj) [inline gdscript2.html.markdown:48]:
```suggestion
var a = [1, false, "brown fox"] # Arrays hold different types of variables at once.
```

File: gdscript2.html.markdown [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+  var a = true
+  var b = false
+  var c = false
+  if a and b or not c: # alternatively you can use &&, || and !
+    print("This is true!")
+
+  for i in range(20): # GDScript's range is similar to Python's
+    print(i) # so this will print numbers from 0 to 19
+
+  for i in 20: # unlike Python, you can loop over an int directly
```
Comment (superherointj) [inline gdscript2.html.markdown:147]:
```suggestion
  for i in 20: # you can loop over an int directly
```

File: gdscript2.html.markdown [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+    print("x and y are equal")
+  
+  var a = true
+  var b = false
+  var c = false
+  if a and b or not c: # alternatively you can use &&, || and !
+    print("This is true!")
+
+  for i in range(20): # GDScript's range is similar to Python's
+    print(i) # so this will print numbers from 0 to 19
```
Comment (superherointj) [inline gdscript2.html.markdown:145]:
```suggestion
  for i in range(20): # prints numbers from 0 to 19
```

File: gdscript2.html.markdown [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+    print("x is smaller than y")
+  elif x > y:
+    print("x is bigger than y")
+  else:
+    print("x and y are equal")
+  
+  var a = true
+  var b = false
+  var c = false
+  if a and b or not c: # alternatively you can use &&, || and !
```
Comment (superherointj) [inline gdscript2.html.markdown:141]:
```suggestion
  if a and b or not c: # alternatively use &&, || and !
```

File: gdscript2.html.markdown [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+    print("This is true!")
+
+  for i in range(20): # GDScript's range is similar to Python's
+    print(i) # so this will print numbers from 0 to 19
+
+  for i in 20: # unlike Python, you can loop over an int directly
+    print(i) # so this will also print numbers from 0 to 19
```
Comment (superherointj) [inline gdscript2.html.markdown:148]:
```suggestion
    print(i) # will print numbers from 0 to 19
```

**Comment by verhovsky** [comment]
We should just update GDScript for Godot 4 and not even mention 3.

@Calinou @Wichamir I changed the PR to overwrite the existing gdscript.html.markdown could you review it please?

**Comment by superherointj** [comment]
Thanks for updating GDScript to Godot 4. Unfortunately my knowledge of GDScript is pretty bad, and when I tried to use Learn X on Y Minutes to learn it, the experience was awful.

**Comment by Calinou** [comment]
> We should just update GDScript for Godot 4 and not even mention 3.

I think it's safe to do this now that we're approaching 4.3, and most newcomers are starting out on 4.x nowadays.

**Comment by verhovsky** [comment]
@Calinou can you review this PR then? Can I merge it?

**Review by Calinou — APPROVED** [review]
Looks good to me.

You could mention support for typed `for` loops (`for i: Node in nodes:`) in a future PR.

**Pull request merged**